### PR TITLE
Bug 1850833 - new telemetry for no-op menu exits

### PR DESCRIPTION
**Pull request by HarrisonOg**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1850833

### CODE REVIEW

**Comment by HarrisonOg** [comment]
DATA REVIEW REQUEST
1. What questions will you answer with this data?

This data will help us understand how users are using the menu.  Specifically, we are looking into the number of times the menu is opened and closed without interaction with any menu item.

2. Why does Mozilla need to answer these questions? Are there benefits for users?
   Do we need this information to address product or business requirements?

We are implementing more telemetry in users' menu activity in order to better understand how we can improve the usefulness of the menu for users.

3. What alternative methods did you consider to answer these questions?
   Why were they not sufficient?

There are no alternative methods to measure menu closing without interaction with menu items or users' interactions with the "Help" button.

4. Can current instrumentation answer these questions?

Currently there is no measurement of users' usage of the "Help" button in the menu or how often users will close the menu without interacting with any menu item.

5. List all proposed measurements and indicate the category of data collection for each
   measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement Name | Measurement Description | Data Collection Category | Tracking Bug
---------------- | ----------------------- | ------------------------ | ------------
`app_menu.user_no_operation_exit` | User has opened and closed the browser menu without interacting with any menu item.  | interaction | https://bugzilla.mozilla.org/show_bug.cgi?id=1850833
`home_menu.help_tapped` | The user clicked the help button in home menu. | interaction | https://bugzilla.mozilla.org/show_bug.cgi?id=1850833
`home_menu.user_no_operation_exit` | User has opened and closed the home page menu without interacting with any menu item.  | interaction | https://bugzilla.mozilla.org/show_bug.cgi?id=1850833

6. Please provide a link to the documentation for this data collection which
   describes the ultimate data set in a public, complete, and accurate way.

This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).

7. How long will this data be collected?

This collection will be collected permanently.
android-probes@mozilla.com, cgordon@mozilla.com will be responsible for the permanent collections.

8. What populations will you measure?

All channels, countries, and locales. No filters.

9. If this data collection is default on, what is the opt-out mechanism for users?

These collections are Glean. The opt-out can be found in the product's preferences.

10. Please provide a general description of how you will analyze this data.

Glean and amplitude

11. Where do you intend to share the results of your analysis?

Glean, Amplitude, and with mobile teams.

12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
    are proposing to use for this data collection?

No.

**Comment by rocketsroger** [comment]
> DATA REVIEW REQUEST
> 
>     1. What questions will you answer with this data?
> 
> 
> This data will help us understand how users are using the menu. Specifically, we are looking into the number of times the menu is opened and closed without interaction with any menu item.
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users?
>        Do we need this information to address product or business requirements?
> 
> 
> We are implementing more telemetry in users' menu activity in order to better understand how we can improve the usefulness of the menu for users.
> 
>     3. What alternative methods did you consider to answer these questions?
>        Why were they not sufficient?
> 
> 
> There are no alternative methods to measure menu closing without interaction with menu items or users' interactions with the "Help" button.
> 
>     4. Can current instrumentation answer these questions?
> 
> 
> Currently there is no measurement of users' usage of the "Help" button in the menu or how often users will close the menu without interacting with any menu item.
> 
>     5. List all proposed measurements and indicate the category of data collection for each
>        measurement, using the Firefox data collection categories found on the Mozilla wiki.
> 
> 
> Measurement Name 	Measurement Description 	Data Collection Category 	Tracking Bug
> `app_menu.user_no_operation_exit` 	User has opened and closed the browser menu without interacting with any menu item. 	interaction 	https://bugzilla.mozilla.org/show_bug.cgi?id=1850833
> `home_menu.help_tapped` 	The user clicked the help button in home menu. 	interaction 	https://bugzilla.mozilla.org/show_bug.cgi?id=1850833
> `home_menu.user_no_operation_exit` 	User has opened and closed the home page menu without interacting with any menu item. 	interaction 	https://bugzilla.mozilla.org/show_bug.cgi?id=1850833
> 
>     6. Please provide a link to the documentation for this data collection which
>        describes the ultimate data set in a public, complete, and accurate way.
> 
> 
> This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).
> 
>     7. How long will this data be collected?
> 
> 
> This collection will be collected permanently. [android-probes@mozilla.com](mailto:android-probes@mozilla.com), [cgordon@mozilla.com](mailto:cgordon@mozilla.com) will be responsible for the permanent collections.
> 
>     8. What populations will you measure?
> 
> 
> All channels, countries, and locales. No filters.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
> These collections are Glean. The opt-out can be found in the product's preferences.
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
> Glean and amplitude
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
> Glean, Amplitude, and with mobile teams.
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
>         are proposing to use for this data collection?
> 
> 
> No.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/apps/fenix )

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, through the "Send Usage Data" preference in the application settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

cgordon@mozilla.com will be responsible for monitoring the data over time

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction data

5) Is the data collection request for default-on or default-off?

default-on

6) Does the instrumentation include the addition of **any *new* identifiers**?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

## Result

data-review+

File: fenix/app/metrics.yaml [hunk 1]
```diff
@@ -9210,6 +9210,23 @@ android_autofill:
...
+    description: The user clicked the help button in home menu.
+    bugs:
+      - https://bugzilla.mozilla.org/show_bug.cgi?id=1850833
+    data_reviews:
+      - https://github.com/mozilla-mobile/firefox-android/pull/3669
+    data_sensitivity:
+      - interaction
+    notification_emails:
+      - android-probes@mozilla.com
+      - cgordon@mozilla.com
```
Comment (rocketsroger) [inline fenix/app/metrics.yaml:9242]: The new process will not require a person's email.  Can you please remove the line with cgordon@mozilla.com.  Thanks

**Review by rocketsroger — COMMENTED** [review]
Looks good, one small change needed.

File: fenix/app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt [hunk 2]
```diff
@@ -111,6 +116,7 @@ class HomeMenu(
         label = context.getString(R.string.browser_menu_desktop_site),
         initialState = { context.settings().openNextTabInDesktopMode },
     ) { checked ->
+        context.metrics.track(Event.UsageData.InteractedWithMenu)
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:119]: Wondering could we do all these track events in https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/home/HomeMenuView.kt#91? We can also add tests for this in https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/test/java/org/mozilla/fenix/home/HomeMenuViewTest.kt

Comment (HarrisonOg) [reply fenix/app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt]: Moved the tracking event to `HomeMenuView` and fixed the corresponding broken tests.

Working on adding new tests to support telemetry.

File: android-components/components/browser/toolbar/src/test/java/mozilla/components/browser/toolbar/display/DisplayToolbarTest.kt [hunk 3]
```diff
@@ -713,9 +713,9 @@ class DisplayToolbarTest {
 
             menuView.impl.performClick()
 
-            assertEquals(1, facts.size)
+            assertEquals(2, facts.size)
```
Comment (rocketsroger) [inline android-components/components/browser/toolbar/src/test/java/mozilla/components/browser/toolbar/display/DisplayToolbarTest.kt:716]: ⚠️ Is this a problem? does this mean one click creates two facts?

**Comment by HarrisonOg** [comment]
Closing this request as this method of determining "no-operation menu exits" is too costly.

**Pull request closed without merging**